[gitlab-use-module] Use gitlab python module instead of raw http requests

CelianR opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

Replaces the gitlab rest api calls by python-gitlab calls

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

- The only business logic difference after refactoring is that `tasks.pipeline.wait_for_pipeline_from_ref` takes the last updated pipeline instead of the last created one
- The gitlab module is now named gitlab_api to avoid conflicts with the package
- The package uses gitlab api v4
- Now the gitlab_api module provides only tokens and function to get the gitlab api / repository object
- To have all items from the api `per_page=100` is the most efficient (max 100 items per page)
- To access functions from objects obtained by `list()`, we need to get the full object using `get(..., lazy=True)`, lazy avoid doing another api call

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

- Each updated task / function has been tested by hand

CelianR committed:
[gitlab-use-module] Added requirements

CelianR committed:
[gitlab-use-module] Added test

CelianR committed:
[gitlab-use-module] Reformatted kmt

CelianR committed:
[gitlab-use-module] Reformatted pipeline data

CelianR committed:
[gitlab-use-module] Tested and reformatted more functions

CelianR committed:
[gitlab-use-module] Tested and reformatted schedule functions

CelianR committed:
[gitlab-use-module] Tested and refactored pipelines + release

CelianR committed:
[gitlab-use-module] Tested and refactored failed jobs

CelianR committed:
[gitlab-use-module] Tested and refactored trigger child pipeline

CelianR committed:
[gitlab-use-module] Tested and refactored kmt

usamasaqib committed:
fix job names in kmt task

CelianR committed:
[gitlab-use-module] Cleaned code, removed remoteapi

CelianR committed:
[gitlab-use-module] Cleaned code, removed tests and todos

CelianR committed:
[gitlab-use-module] Updated pipeline_notifications with new job object class

CelianR committed:
[gitlab-use-module] Removed TODOs

CelianR committed:
[gitlab-use-module] Cleaned code

CelianR committed:
[gitlab-use-module] Cleaned code

CelianR committed:
[gitlab-use-module] Fixed release.py refactoring

KevinFairise2 reviewed (approved):
LGTM! I did not check that all the methods we use on Gitlab objects return exactly what is expected, but that looks good.

Good job 🎉

chouetz commented on tasks/kmt.py:
```diff
@@ -127,37 +127,35 @@ def gen_config_from_ci_pipeline(
     Generate a vmconfig.json file with the VMs that failed jobs in the given pipeline.
     """
-    gitlab = Gitlab("DataDog/datadog-agent", str(get_gitlab_token()))
+    repo = get_gitlab_repo()
```
nitpick (just to illustrate it's this one you get by default)
```suggestion
    datadog_agent = get_gitlab_repo()
```

chouetz commented on tasks/kmt.py:
```diff
@@ -127,37 +127,35 @@ def gen_config_from_ci_pipeline(
 
         if (
             (vcpu is None or memory is None)
             and name.startswith("kernel_matrix_testing_setup_env")
```
I can confirm the [names have changed](https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/472524203)
```suggestion
            and name.startswith("kmt_setup_env")
```

CelianR replied:
It has already been made in this PR : https://github.com/DataDog/datadog-agent/pull/24179

chouetz commented on tasks/kmt.py:
```diff
@@ -168,7 +166,7 @@ def gen_config_from_ci_pipeline(
                     vcpu = str(vcpu_list[0])
                     info(f"[+] setting vcpu to {vcpu}")
-        elif name.startswith("kernel_matrix_testing_run") and job["status"] == "failed":
+        elif name.startswith("kernel_matrix_testing_run") and job.status == "failed":
```
```suggestion
        elif name.startswith("kmt_run") and job.status == "failed":
```

CelianR committed:
[gitlab-use-module] Removed requirements (moved to buildimages)

chouetz commented on tasks/libs/pipeline_data.py:
```diff
@@ -1,54 +1,56 @@
     """
     Retrieves the list of failed jobs for a given pipeline id in a given project.
     """
+    repo = get_gitlab_repo(project_name)
+    pipeline = repo.pipelines.get(pipeline_id)
```
```suggestion
    datadog_agent = get_gitlab_repo(project_name)
    pipeline = datadog_agent.pipelines.get(pipeline_id)
```

chouetz commented on tasks/libs/pipeline_notifications.py:
```diff
@@ -56,13 +58,16 @@ def check_for_missing_owners_slack_and_jira(print_missing_teams=True, owners_fil
-def get_failed_tests(project_name, job, owners_file=".github/CODEOWNERS"):
-    gitlab = Gitlab(project_name=project_name, api_token=get_gitlab_token())
+def get_failed_tests(project_name, job: ProjectJob, owners_file=".github/CODEOWNERS"):
+    repo = get_gitlab_repo(project_name)
```
```suggestion
    datadog_agent = get_gitlab_repo(project_name)
```

chouetz commented on tasks/libs/pipeline_notifications.py:
```diff
@@ -56,13 +58,16 @@ def check_for_missing_owners_slack_and_jira(print_missing_teams=True, owners_fil
     owners = read_owners(owners_file)
-    test_output = gitlab.artifact(job["id"], "test_output.json", ignore_not_found=True)
+    try:
+        test_output = str(repo.jobs.get(job.id, lazy=True).artifact('test_output.json'), 'utf-8')
```
Not related to your change, but I'm wondering why we don't `json.loads` the `test_output.json` right away but line by line below.

CelianR replied:
This file is not really a json file but contains json for each line, e.g. :
```json
{"hello": "world"}
{"cats": [...]}
```

chouetz commented on tasks/release.py:
```diff
@@ -1328,7 +1329,7 @@ def build_rc(ctx, major_versions="6,7", patch_version=False, k8s_deployments=Fal
         return Exit(message="Must use Python 3 for this task", code=1)
 
-    gitlab = Gitlab(project_name=GITHUB_REPO_NAME, api_token=get_gitlab_token())
+    repo = get_gitlab_repo()
```
```suggestion
    datadog_agent = get_gitlab_repo()
```
GITHUB_REPO_NAME might be unused?

chouetz reviewed (commented):
Not related to your change, we probably can take action on the `kernel_matrix_testing` names.

CelianR committed:
[gitlab-use-module] Applied suggestions

CelianR committed:
[gitlab-use-module] Renammed gitlab repo variable

CelianR committed:
[gitlab-use-module] Pinned another build images version

CelianR committed:
[gitlab-use-module] Cleaned code

CelianR committed:
[gitlab-use-module] Added requirements to assign-team-label

CelianR committed:
[gitlab-use-module] Updated find job owner

CelianR committed:
[gitlab-use-module] Updated notify

CelianR committed:
[gitlab-use-module] Updated pipeline_data, cleaned code

CelianR committed:
[gitlab-use-module] Convert FailedJobType (fixed error)

CelianR committed:
[gitlab-use-module] Updated requirements command, fixed error

CelianR committed:
[gitlab-use-module] Updated build image

gjulianm reviewed (approved):
LGTM for ebpf-platform files, thanks!

CelianR committed:
[gitlab-use-module] Test ci system probe

CelianR committed:
[gitlab-use-module] Test ci system probe

CelianR committed:
[gitlab-use-module] Fixed notify tests

CelianR committed:
[update-buildimages] Updated build images version

CelianR committed:
[gitlab-use-module] Fixed notify tests with mocks

CelianR committed:
[update-buildimages] Source bashrc

CelianR committed:
[update-buildimages] Source bashrc

[update-buildimages] Added source /root/.bashrc to sysprobe scripts

[update-buildimages] Added source /root/.bashrc to extra jobs

[update-buildimages] Added source /root/.bashrc to before_script

[update-buildimages] Added remaining source /root/.bashrc to before_script

CelianR committed:
[update-buildimages] Used v31342166-4135b1c3

CelianR committed:
[update-buildimages] Used v31342166-4135b1c3

CelianR committed:
[update-buildimages] Added source /root/.bashrc to sysprobe scripts

CelianR committed:
[update-buildimages] Added source /root/.bashrc to extra jobs

CelianR committed:
[update-buildimages] Added source /root/.bashrc to before_script

CelianR committed:
[update-buildimages] Added remaining source /root/.bashrc to before_script

CelianR committed:
[gitlab-use-module] Merged + refactored gitlab code

CelianR committed:
[gitlab-use-module] Sourced bashrc

CelianR committed:
[gitlab-use-module] Updated build images

CelianR commented:
/trigger-ci

CelianR committed:
[update-buildimages] Updated version

[update-buildimages] Updated version

CelianR committed:
[gitlab-use-module] Removed added source /root/.bashrc

CelianR committed:
Revert "[update-buildimages] Source bashrc"

This reverts commit bd5b3a74be24703e26e46e051495fe23633c0b1f.

CelianR committed:
[gitlab-use-module] Updated build images

CelianR commented:
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

CelianR committed:
Merged with main for new build images

CelianR committed:
kmt: Refactoring fixes

CelianR committed:
[gitlab-use-module] Fixed notify tests

CelianR committed:
Merge branch 'main' into celian/gitlab-use-module-acix-65

CelianR committed:
[gitlab-use-module] Sync with main

CelianR committed:
[gitlab-use-module] Updated test infra version

CelianR committed:
[gitlab-use-module] Merged main, refactored gitlab ci lint

CelianR committed:
[gitlab-use-module] Merged main

CelianR committed:
[gitlab-use-module] Updated notify_test with good trace

chouetz commented on tasks/pipeline.py:
```diff
@@ -582,21 +579,16 @@ def changelog(ctx, new_commit_sha):
+    repo = get_gitlab_repo(token=get_gitlab_bot_token())
+
+    for sched in repo.pipelineschedules.list(per_page=100, all=True):
+        sched.pprint()
```
Is there really a `pprint` method in the returned object of `pipelineschedules` (so [this one](https://python-gitlab.readthedocs.io/en/stable/api/gitlab.v4.html#gitlab.v4.objects.ProjectPipelineSchedule) I guess)?
Or I misunderstood something?
Same remark on the chunks below.
```suggestion
        pprint.pprint(sched)
```
Even if I'm not sure what `pprint.pprint` will do in that kind of object

chouetz commented on tasks/pipeline.py:
```diff
@@ -629,9 +627,14 @@ def edit_schedule(_, schedule_id, description=None, ref=None, cron=None, cron_ti
+
+    sched = repo.pipelineschedules.update(schedule_id, data)
+
+    pprint.pprint(sched)
```
HA 😄 
But I'm not sure about what it would output (if it's similar to what it was).

chouetz commented on tasks/pipeline.py:
```diff
@@ -662,9 +669,11 @@ def delete_schedule(_, schedule_id):
+
+    repo.pipelineschedules.delete(schedule_id)
+
+    print('Deleted schedule', schedule_id)
```
This one is yet another solution

chouetz commented on tasks/pipeline.py:
```diff
@@ -908,28 +926,28 @@ def test_merge_queue(ctx):
     # Search for the generated pipeline
     print(f"PR {pr.html_url} is waiting for MQ pipeline generation")
-    gitlab = Gitlab(api_token=get_gitlab_token())
+    gitlab = get_gitlab_repo()
```
```suggestion
    agent = get_gitlab_repo()
```

chouetz reviewed (commented):
I'd like a check on the `pprint.pprint` part at least, thanks in advance

CelianR commented on tasks/libs/pipeline/tools.py:
```diff
@@ -32,55 +36,55 @@ def parse_datetime(dt):
             f"{color_message('Started at', 'blue')} {parse_datetime(pipeline_creation_date).astimezone():%c} ({pipeline_creation_date})"
         )
 
         print(
```
It cannot be replaced by such function since it dumps all attributes and it is too much verbose

CelianR committed:
[gitlab-use-module] Applied suggestions, cleaned code

chouetz commented on tasks/libs/pipeline/tools.py:
```diff
@@ -307,26 +307,22 @@ def print_job(name, stage, color, date, duration, status, link):
     # A None duration is set by Gitlab when the job gets canceled before it was started.
     # In that case, set a duration of 0s.
-    if duration is None:
+    if job.finished_at is None:
```
```suggestion
    if job.duration is None:
```

chouetz reviewed (commented):
I think there is a last change needed, thanks a lot

CelianR committed:
[gitlab-use-module] Cleaned code

chouetz reviewed (approved):
ship it!
